src/cryptonote_core/miner.cpp

```diff
@@ -411,5 +449,294 @@ namespace cryptonote
+    uint64_t total_sleep_time = 0;
+
+    uint64_t prev_total_time, current_total_time;
+    uint64_t prev_idle_time, current_idle_time;
+    uint64_t previous_process_time = 0, current_process_time = 0;
+
+    // get some starting data, then wait one windows
+    if(!get_system_times(prev_total_time, prev_idle_time))
+    {
+      LOG_ERROR(__func__ << " : get_system_times call failed, background mining will NOT work!");
+      return false;
+    }
+
+    while(!m_stop)
+    {
+      misc_utils::sleep_no_w(SLEEP_CYCLE); // zzZZzzz
+      if( !m_is_background_mining_enabled ) continue;
+      total_sleep_time += SLEEP_CYCLE; // at least ..
+
+      if( m_is_background_mining_started )
+      {
+        // figure out if we need to stop, and monitor mining usage
+
+        // previous is uninitialized
+        if(previous_process_time == 0)
+        {
+          // get some starting data, then wait one windows
+          if(!get_system_times(prev_total_time, prev_idle_time))
+            LOG_ERROR(__func__ << " : get_system_times call failed");
+
+          if(!get_process_time(previous_process_time))
+            LOG_ERROR(__func__ << " : get_process_time call failed!");
+
+          continue;
+        }
+
+        // has enough time passed for analysis?
+        if( total_sleep_time < BACKGROUND_MINING_MINER_MONITOR_INVERVAL_IN_SECONDS * 1000 )
+          continue;
+
+        // yay, if we're here then enough time has passed since our last reading!
+
+        if(!get_system_times(current_total_time, current_idle_time))
+        {
+          LOG_ERROR(__func__ << " : get_system_times call failed");
+          continue;
+        }
+
+        if(!get_process_time(current_process_time))
+        {
+          LOG_ERROR(__func__ << " : get_process_time call failed!");
+          continue;
+        }
 
+        uint64_t total_diff = (current_total_time - prev_total_time);
+        uint64_t idle_diff = (current_idle_time - prev_idle_time);
+        uint64_t process_diff = (current_process_time - previous_process_time);
+
+        // using ceil here is giving me a little edge, but just be a good network citizen and mine!
+        uint8_t idle_percentage = (uint8_t)( ceil( (idle_diff * 1.f / total_diff * 1.f) * 100) );
+        uint8_t process_percentage = (uint8_t)( ceil( (process_diff * 1.f / total_diff * 1.f) * 100) );
+        MGINFO(__func__ << " : idle percentage is " << unsigned(idle_percentage) << "\%, miner percentage is " << unsigned(process_percentage) << "\%");
+
+        if( idle_percentage + process_percentage < get_idle_threshold() )
+        {
+          MGINFO(__func__ << " : cpu is " << unsigned(idle_percentage) << "% idle, idle threshold is " << unsigned(get_idle_threshold()) << "\%, background mining stopping, thanks for your contribution!");
+          m_is_background_mining_started = false;
+
+          // reset process times
+          previous_process_time = 0;
```

Comment:
Why is the sleep interval at the top of this loop for a shorter time period than the entire `MONITOR_INTERVAL`? Why is the logic to explicitly wait one short sleep cycle before grabbing the initial value for `previous_process_time`?

The way this is currently written, it appears the goal is to wait a short interval _after_ enabling background mining before setting the `previous_process_time` value. Look at every other branch - if the entire `MONITOR_INTERVAL` has not elapsed the loop is `continue`d to sleep a short time interval again. Only one branch does anything different, and it is the case where `previous_process_time` is set for the first time after mining is enabled.

Is the intent to wait after enabling mining before setting `previous_process_time`? Or ...? What I am suggesting is basically:

```c++
while (!m_stop) {
    boost::this_thread::sleep_for(BACKGROUND_MINING_MONITOR_INTERVAL);
    if (m_is_background_mining_started) {
        if (percentage(...) + percentage(...) < get_idle_threshold()) {
            m_is_background_mining_started = false;
            ...
        } else {
            ...
        }
    } else {
        if (get_idle_threshold() <= percentage(...)) {
            m_is_background_mining_started = true;
            boost::this_thread::sleep_for(std::chrono::milliseconds(100));
            // this is the only case where something is done between MONITOR_INTERVALs
            ...
            if (!get_process_time(previous_process_time))
            ...
        } else {
            ...
        }
    }
}
```